# **Numpy**

In [2]:
import numpy as np

Essa lista contém questões práticas incluindo os seguintes tópicos com numpy.

* Criação de array e seus atributos, intervalos numéricos em numPy
* Slicing e indexação de NumPy Array.
* Manipulação, pesquisa, classificação e divisão de arrays.
* Funções matemáticas com arrays e broadcasting arrays NumPy. 

## **Crie um array inteiro 4x2 e mostre seus atributos**

Os elementos devem ser um tipo de int16 sem sinal. E imprima os seguintes atributos: formato do array; dimensões; comprimento de cada elemento do array.

In [7]:
array = np.zeros((4, 2), dtype=np.uint16)

print("Array:")
print(array)

print("Formato do Array:", array.shape)
print("Número de dimensões:", array.ndim)
print("Comprimento de Cada Elemento do Array:", array.itemsize, "bytes")

Array:
[[0 0]
 [0 0]
 [0 0]
 [0 0]]
Formato do Array: (4, 2)
Número de dimensões: 2
Comprimento de Cada Elemento do Array: 2 bytes


##### Resultado esperado:
<pre>
Mostrando Array
[[64392 31655]
 [32579     0]
 [49248   462]
 [    0     0]]

Mostrando Attributos:

Formato do array:  (4, 2)
Número de dimensões:  2
Comprimento de cada elemento do array:  2
</pre>

## **Crie uma array de inteiros 5X2 com um intervalo entre 100 e 200, de modo que a diferença entre cada elemento seja 10**

In [6]:
array = np.arange(100, 200, 10).reshape(5, 2)

print(array)

[[100 110]
 [120 130]
 [140 150]
 [160 170]
 [180 190]]


##### Resultado esperado:
<pre>
[[100 110]
 [120 130]
 [140 150]
 [160 170]
 [180 190]]
</pre>

## **Baseado no array fornecido. Retorne um array de itens com a terceira coluna de todas as linhas**

In [9]:
array_00 = np.array([[11 ,22, 33], [44, 55, 66], [77, 88, 99]])

In [10]:
third_column = array_00[:, 2]
print(third_column)

[33 66 99]


##### Resultado esperado:
<pre>
Mostrando array de entrada
[[11 22 33]
 [44 55 66]
 [77 88 99]]

Novo array com a terceira coluna de cada linha
[33 66 99]
</pre>

## **Retorne um array com os valores presentes nas linhas ímpares com colunas pares baseado no seguinte array**

In [11]:
array_01 = np.array([[3 ,6, 9, 12], [15 ,18, 21, 24], 
[27 ,30, 33, 36], [39 ,42, 45, 48], [51 ,54, 57, 60]])

In [14]:
odd_rows_even_cols = array_01[0::2, 1::2]
print(odd_rows_even_cols)


[[ 6 12]
 [30 36]
 [54 60]]


##### Resultado esperado:
<pre>
Mostrando array de entrada
[[ 3  6  9 12]
 [15 18 21 24]
 [27 30 33 36]
 [39 42 45 48]
 [51 54 57 60]]

Array resultante (linha ímpar x coluna par)
[[ 6 12]
 [30 36]
 [54 60]]
</pre>

## **Crie um array de resultado somando os seguintes arrays. Em seguida, modifique o array resultante calculando o quadrado de cada elemento**

In [19]:
array_02 = np.array([[5, 6, 9], [21 ,18, 27]])
array_03 = np.array([[15 ,33, 24], [4 ,7, 1]])

In [23]:
result = array_02 + array_03
print("Soma:")
print(result)

result = np.square(result)
print("Quadrado:")
print(result)

Soma:
[[20 39 33]
 [25 25 28]]
Quadrado:
[[ 400 1521 1089]
 [ 625  625  784]]


##### Resultado esperado:
<pre>
A soma dos arrays é: 
[[20 39 33]
 [25 25 28]]

Resultado depois de calcular o quadrado de cada número:
[[ 400 1521 1089]
 [ 625  625  784]]
</pre>

## **Faça a soma, subtração, multiplicação e divisão dos arrays**

In [25]:
array_04 = np.array([1, 2, 3])
array_05 = np.array([[1, 2, 3], [4, 5, 6]])

In [26]:
sum = array_04 + array_05
print("Soma:")
print(sum)

sub = array_04 - array_05
print("Subtração:")
print(sub)

mul = array_04 * array_05
print("Multiplicação:")
print(mul)

div = array_04 / array_05
print("Divisão:")
print(div)

Soma:
[[2 4 6]
 [5 7 9]]
Subtração:
[[ 0  0  0]
 [-3 -3 -3]]
Multiplicação:
[[ 1  4  9]
 [ 4 10 18]]
Divisão:
[[1.   1.   1.  ]
 [0.25 0.4  0.5 ]]


##### Resultado esperado:
<pre>
Soma: 
[[2 4 6]
 [5 7 9]]
Subtração: 
[[ 0  0  0]
 [-3 -3 -3]]
Multiplicação: 
[[ 1  4  9]
 [ 4 10 18]]
divisão: 
[[1.   1.   1.  ]
 [0.25 0.4  0.5 ]]
</pre>

## **Divida o array em quatro sub-arrays de tamanhos iguais**

Crie um array de inteiros 8X3 com um intervalo entre 10 e 34 de modo que a diferença entre cada elemento seja 1 e, em seguida, faça a subdivisão. 

In [36]:
array = np.arange(10, 34, 1).reshape(8, 3)
print(array)

sub_arrays = np.split(array, 4)

for i, sub_array in enumerate(sub_arrays):
    print(f"array {i+1}:")
    print(sub_array)

[[10 11 12]
 [13 14 15]
 [16 17 18]
 [19 20 21]
 [22 23 24]
 [25 26 27]
 [28 29 30]
 [31 32 33]]
array 1:
[[10 11 12]
 [13 14 15]]
array 2:
[[16 17 18]
 [19 20 21]]
array 3:
[[22 23 24]
 [25 26 27]]
array 4:
[[28 29 30]
 [31 32 33]]


##### Resultado esperado:
<pre>
Criando o array de 8X3 usando numpy.arange
[[10 11 12]
 [13 14 15]
 [16 17 18]
 [19 20 21]
 [22 23 24]
 [25 26 27]
 [28 29 30]
 [31 32 33]]

Dividindo o array em 4 sub-arrays

[array([[10, 11, 12],[13, 14, 15]]), 
array([[16, 17, 18],[19, 20, 21]]), 
array([[22, 23, 24],[25, 26, 27]]), 
array([[28, 29, 30],[31, 32, 33]])]
</pre>

## **Ordenar o seguinte array da seguinte forma:**

* Ordene pela segunda linha
* Ordene pela segunda coluna

In [37]:
array_06 = np.array([[34,43,73],[82,22,12],[53,94,66]])

In [43]:
indices_ordenados = np.argsort(array_06[1])

array = array_06[:, indices_ordenados]
print("Segunda linha:")
print(array)

indices_ordenados = np.argsort(array_06[:, 2])

array = array_06[:, indices_ordenados[::-1]]
print("Terceira coluna:")
print(array)

Segunda linha:
[[73 43 34]
 [12 22 82]
 [66 94 53]]
Terceira coluna:
[[34 73 43]
 [82 12 22]
 [53 66 94]]


##### Resultado esperado:
<pre>
Mostrando o array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Ordenando o array pela segunda linha
[[73 43 34]
 [12 22 82]
 [66 94 53]]

Ordenando o array pela terceira coluna
[[82 22 12]
 [34 43 73]
 [53 94 66]]
</pre>

## **Imprima o máx. do eixo 0 e mín. Do eixo 1 do seguinte array**

In [45]:
array_07 = np.array([[34,43,73],[82,22,12],[53,94,66]])

In [46]:
max_0 = np.amax(array_07, axis=0)
print("Máximo por coluna:")
print(max_0)

min_1 = np.amin(array_07, axis=1)
print("Mínimo por linha:")
print(min_1)

Máximo por coluna:
[82 94 73]
Mínimo por linha:
[34 12 53]


##### Resultado esperado:
<pre>
Mostrando array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Mostando amin do eixo 1
[34 12 53]

Mostrando amax do eixo 0
[82 94 73]
</pre>

## **Exclua a segunda coluna de um determinado array e insira a nova coluna seguinte em seu lugar**

In [48]:
array_08    = np.array([[34,43,73],[82,22,12],[53,94,66]]) 
nova_coluna = np.array([[10,10,10]])

In [54]:
array = np.delete(array_08, 1, axis=1)
print(array)

array = np.insert(array, 1, nova_coluna, axis=1)
print("Inserindo")
print(array)

[[34 73]
 [82 12]
 [53 66]]
Inserindo
[[34 10 73]
 [82 10 12]
 [53 10 66]]


##### Resultado esperado:
<pre>
Mostrando array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Array depois de deletar a segunda coluna
[[34 73]
 [82 12]
 [53 66]]

Array depois de inserir a nova coluna no lugar da coluna excluída
[[34 10 73]
 [82 10 12]
 [53 10 66]]
</pre>

# Pandas

### 1. Fazer o download dos dados do desafio https://www.kaggle.com/t/5e53165f23784ce59e4379f9274e1af4

### 2. Carregar os dados no Pandas.

### 3. Usar esses dados para praticar os comandos básicos mostrados sobre Pandas. Procure extrapolar o que foi passado em aula, visando descobrir outras funcionalidades da biblioteca. 